c-deps/libroach/db.cc

```diff
@@ -697,9 +697,36 @@ DBStatus DBIngestExternalFiles(DBEngine* db, char** paths, size_t len, bool move
   ingest_options.allow_global_seqno = allow_file_modifications;
   // If there are mutations in the memtable for the keyrange covered by the file
   // being ingested, this option is checked. If true, the memtable is flushed
-  // and the ingest run. If false, an error is returned.
-  ingest_options.allow_blocking_flush = true;
+  // using a blocking, write-stalling flush and the ingest run. If false, an
+  // error is returned.
+  // We want to ingest, but we do not want a write-stall, so we initially set it
+  // to false -- if our ingest fails, we'll do a manual, no-stall flush and wait
+  // for it to finish before trying the ingest again.
+  ingest_options.allow_blocking_flush = false;
+
   rocksdb::Status status = db->rep->IngestExternalFile(paths_vec, ingest_options);
+  if (status.IsInvalidArgument()) {
+    // TODO(dt): inspect status to see if it has the message `External file requires flush`
+    //           since the move_file and other errors also use kInvalidArgument.
+
+    // It is possible we failed because the memtable wanted to flush but we did
+    // not allow a blocking flush on the first try. Do a manual, non-blocking
+    // flush and wait for it, then try again.
+    rocksdb::FlushOptions flush_options;
+    flush_options.allow_write_stall = false;
```

Comment:
Note that prior to https://github.com/cockroachdb/rocksdb/pull/25, the flush performed by `IngestExternalFile` was being called with `allow_write_stall == false`. This was a bug (and caused deadlock). My point is that this change shouldn't have an effect until that PR is merged and we bump the RocksDB pointer.

Reply:
Yep, agreed that it won't have a p99 effect until that fix lands.

Though I think on current rocks code, this change means we hit the deadlock situation less often -- I think we don't even try to do the deadlocking flush when `allow_blocking_flush` is false, then we do the flush explicitly and unless we get unlucky on the retry, we don't need a flush that time and dodge the deadlock?

Reply:
https://github.com/cockroachdb/rocksdb/pull/25 landed and the rocksdb pointer has been bumped. This change should be a nice complement now.